phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLog.java

```diff
@@ -0,0 +1,908 @@
+        LOG.info("ReplicationLogWriter started with ring buffer size {}", ringBufferSize);
+    }
+
+    /**
+     * Append a mutation to the log. This method is non-blocking and returns quickly, unless the
+     * ring buffer is full. The actual write happens asynchronously. We expect multiple append()
+     * calls followed by a sync(). The appends will be batched by the Disruptor. Should the ring
+     * buffer become full, which is not expected under normal operation but could (and should)
+     * happen if the log file writer is unable to make progress, due to a HDFS level disruption.
+     * Should we enter that condition this method will block until the append can be inserted.
+     * @param tableName The name of the HBase table the mutation applies to.
+     * @param commitId  The commit identifier (e.g., SCN) associated with the mutation.
+     * @param mutation  The HBase Mutation (Put or Delete) to be logged.
+     * @throws IOException If the writer is closed or if the ring buffer is full.
+     */
+    public void append(String tableName, long commitId, Mutation mutation) throws IOException {
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Append: table={}, commitId={}, mutation={}", tableName, commitId, mutation);
+        }
+        if (isClosed) {
+            throw new IOException("Closed");
+        }
+        long startTime = System.nanoTime();
+        // ringBuffer.next() claims the next sequence number. Because we initialize the Disruptor
+        // with ProducerType.MULTI and the blocking YieldingWaitStrategy this call WILL BLOCK if
+        // the ring buffer is full, thus providing backpressure to the callers.
+        long sequence = ringBuffer.next();
+        try {
+            LogEvent event = ringBuffer.get(sequence);
+            event.setValues(EVENT_TYPE_DATA, new Record(tableName, commitId, mutation), null,
+                sequence);
+            metrics.updateAppendTime(System.nanoTime() - startTime);
+        } finally {
+            // Update ring buffer events metric
+            ringBuffer.publish(sequence);
+        }
+    }
+
+    /**
+     * Ensures all previously appended records are durably persisted. This method blocks until the
+     * sync operation completes or fails, potentially after internal retries. All in flight appends
+     * are batched and provided to the underlying LogWriter, which will then be synced. If there is
+     * a problem syncing the LogWriter we will retry, up to the retry limit, rolling the writer for
+     * each retry.
+     * <p>
+     * NOTE: When the ReplicationLogManager is capable of switching between synchronous and
+     * fallback (store-and-forward) writers, then this will be pretty bullet proof. Right now we
+     * will still try to roll the synchronous writer a few times before giving up.
+     * @throws IOException If the sync operation fails after retries, or if interrupted.
+     */
+    public void sync() throws IOException {
```

Comment:
What will be the behavior of sync when multiple rpc handler threads will do sync in parallel on the same underlying ReplicationLog ?

Reply:
If I understand this correctly, we will publish a SYNC event for every sync call by the rpc handler and will do a filesystem sync for every SYNC event. Moreover, each SYNC will also create a new block

Reply:
Yes, unless there are no pending in-flight appends. I think it has to be this way. Any caller of sync expects when the sync returns that all of its in-flight appends have been persisted and also the sync has been propagated through to the underlying filesystem. 

My focus now is on correctness and simplicity in implementation is part of that. If the callers have a requirement that sync calls are optimized, this concern for now is pushed to the callers. I do think there are opportunities to do sync merging/consolidation, like how HBase WAL does, as a follow on.

Reply:
Addressed in 2a3b0f620